File: services/accounts/accounts.go
package accounts

import (
	"context"
	"errors"
	"strings"

	"github.com/ethereum/go-ethereum/common"

	accsmanagement "github.com/status-im/status-go/internal/accounts-management"
	accscommon "github.com/status-im/status-go/internal/accounts-management/common"
	"github.com/status-im/status-go/internal/accounts-management/generator"
	accsmanagementtypes "github.com/status-im/status-go/internal/accounts-management/types"
	"github.com/status-im/status-go/internal/crypto/types"
	"github.com/status-im/status-go/internal/db/multiaccounts/accounts"
	walletsettings "github.com/status-im/status-go/internal/db/multiaccounts/settings_wallet"
	"github.com/status-im/status-go/internal/logutils"
	"github.com/status-im/status-go/internal/protocol"
	"github.com/status-im/status-go/params"
	"github.com/status-im/status-go/pkg/pubsub"
	"github.com/status-im/status-go/services/accounts/accountsevent"
)

func NewAccountsAPI(manager *accsmanagement.AccountsManager, config *params.NodeConfig, db *accounts.Database, messenger **protocol.Messenger, publisher *pubsub.Publisher) *API {
	return &API{manager, config, db, messenger, publisher}
}

// API is class with methods available over RPC.
type API struct {
	manager   *accsmanagement.AccountsManager
	config    *params.NodeConfig
	db        *accounts.Database
	messenger **protocol.Messenger
	publisher *pubsub.Publisher
}

type DerivedAddress struct {
	Address        common.Address `json:"address"`
	Path           string         `json:"path"`
	HasActivity    bool           `json:"hasActivity"`
	AlreadyCreated bool           `json:"alreadyCreated"`
}

func (api *API) publishAccountsEvent(accounts []*accsmanagementtypes.Account, removeEvent bool) {
	if api.publisher != nil {
		commonAddresses := []common.Address{}
		for _, acc := range accounts {
			commonAddresses = append(commonAddresses, common.Address(acc.Address))
		}

		if removeEvent {
			pubsub.Publish(api.publisher, accountsevent.AccountsRemovedEvent{
				Accounts: commonAddresses,
			})
		} else {
			pubsub.Publish(api.publisher, accountsevent.AccountsAddedEvent{
				Accounts: commonAddresses,
			})
		}
	}
}

func (api *API) AddKeypairViaSeedPhrase(ctx context.Context, mnemonic string, password string, name string,
	coldWallet accsmanagementtypes.ColdWalletType, walletAccount *accsmanagementtypes.AccountCreationDetails) (*accsmanagementtypes.Keypair, error) {

	addedKeypair, err := (*api.messenger).AddKeypairViaSeedPhrase(mnemonic, password, name, coldWallet, walletAccount)
	if err != nil {
		return nil, err
	}

	api.publishAccountsEvent(addedKeypair.Accounts, false)

	return addedKeypair, nil
}

func (api *API) AddKeypairStoredToColdWallet(ctx context.Context, keyUID string, masterAddress string, name string,
	walletXPub string, coldWallet accsmanagementtypes.ColdWalletType, walletAccounts []*accsmanagementtypes.Account) (*accsmanagementtypes.Keypair, error) {
	addedKeypair, err := (*api.messenger).AddKeypairStoredToColdWallet(keyUID, masterAddress, name, walletXPub, coldWallet, walletAccounts)
	if err != nil {
		return nil, err
	}

	api.publishAccountsEvent(addedKeypair.Accounts, false)

	return addedKeypair, nil
}

func (api *API) AddKeypairViaPrivateKey(ctx context.Context, privateKey string, password string, name string,
	walletAccount *accsmanagementtypes.AccountCreationDetails) (*accsmanagementtypes.Keypair, error) {
	addedKeypair, err := (*api.messenger).AddKeypairViaPrivateKey(privateKey, password, name, walletAccount)
	if err != nil {
		return nil, err
	}

	api.publishAccountsEvent(addedKeypair.Accounts, false)

	return addedKeypair, nil
}

func (api *API) AddAccount(ctx context.Context, password string, account *accsmanagementtypes.Account) error {
	if account.Type == accsmanagementtypes.AccountTypeGenerated {
		account.AddressWasNotShown = true
	}

	err := (*api.messenger).AddAccount(account, password)
	if err != nil {
		return err
	}

	api.publishAccountsEvent([]*accsmanagementtypes.Account{account}, false)

	return nil
}

func (api *API) UpdateAccount(ctx context.Context, account *accsmanagementtypes.Account) error {
	logutils.ZapLogger().Info("[AccountsAPI::SaveAccount]")
	err := (*api.messenger).UpdateAccount(account)
	if err != nil {
		return err
	}

	api.publishAccountsEvent([]*accsmanagementtypes.Account{account}, false)

	return nil
}

// Setting `Keypair` without `Accounts` will update keypair only, `Keycards` won't be saved/updated this way.
func (api *API) UpdateKeypair(ctx context.Context, keypair *accsmanagementtypes.Keypair) error {
	logutils.ZapLogger().Info("[AccountsAPI::SaveKeypair]")
	err := (*api.messenger).UpdateKeypair(keypair)
	if err != nil {
		return err
	}

	api.publishAccountsEvent(keypair.Accounts, false)

	return nil
}

func (api *API) HasPairedDevices(ctx context.Context) bool {
	return (*api.messenger).HasPairedDevices()
}

// Setting `Keypair` without `Accounts` will update keypair only.
func (api *API) UpdateKeypairName(ctx context.Context, keyUID string, name string) error {
	return (*api.messenger).UpdateKeypairName(keyUID, name)
}

func (api *API) MoveWalletAccount(ctx context.Context, fromPosition int64, toPosition int64) error {
	return (*api.messenger).MoveWalletAccount(fromPosition, toPosition)
}

func (api *API) UpdateTokenPreferences(ctx context.Context, preferences []walletsettings.TokenPreferences) error {
	return (*api.messenger).UpdateTokenPreferences(preferences)
}

func (api *API) GetTokenPreferences(ctx context.Context) ([]walletsettings.TokenPreferences, error) {
	return (*api.messenger).GetTokenPreferences()
}

func (api *API) UpdateCollectiblePreferences(ctx context.Context, preferences []walletsettings.CollectiblePreferences) error {
	return (*api.messenger).UpdateCollectiblePreferences(preferences)
}

func (api *API) GetCollectiblePreferences(ctx context.Context) ([]walletsettings.CollectiblePreferences, error) {
	return (*api.messenger).GetCollectiblePreferences()
}

func (api *API) GetAccounts(ctx context.Context) ([]*accsmanagementtypes.Account, error) {
	return api.db.GetActiveAccounts()
}

func (api *API) GetWatchOnlyAccounts(ctx context.Context) ([]*accsmanagementtypes.Account, error) {
	return api.db.GetActiveWatchOnlyAccounts()
}

func (api *API) GetKeypairs(ctx context.Context) ([]*accsmanagementtypes.Keypair, error) {
	return api.db.GetActiveKeypairs()
}

func (api *API) GetAccountByAddress(ctx context.Context, address types.Address) (*accsmanagementtypes.Account, error) {
	return api.db.GetAccountByAddress(address)
}

func (api *API) GetKeypairByKeyUID(ctx context.Context, keyUID string) (*accsmanagementtypes.Keypair, error) {
	return api.db.GetKeypairByKeyUID(keyUID)
}

func (api *API) DeleteAccount(ctx context.Context, address types.Address, password string) error {
	err := (*api.messenger).DeleteAccount(address, password)
	if err != nil {
		return err
	}

	api.publishAccountsEvent([]*accsmanagementtypes.Account{{Address: address}}, true)

	return nil
}

func (api *API) DeleteKeypair(ctx context.Context, keyUID string, password string) error {
	keypair, err := api.db.GetKeypairByKeyUID(keyUID)
	if err != nil {
		return err
	}

	if keypair.Type == accsmanagementtypes.KeypairTypeProfile {
		return accounts.ErrCannotRemoveProfileKeypair
	}

	err = (*api.messenger).DeleteKeypair(keyUID, password)
	if err != nil {
		return err
	}

	api.publishAccountsEvent(keypair.Accounts, true)

	return nil
}

// RemainingAccountCapacity returns the number of accounts that can be added.
func (api *API) RemainingAccountCapacity(ctx context.Context) (int, error) {
	return (*api.messenger).RemainingAccountCapacity()
}

// RemainingKeypairCapacity returns the number of keypairs that can be added.
func (api *API) RemainingKeypairCapacity(ctx context.Context) (int, error) {
	return (*api.messenger).RemainingKeypairCapacity()
}

// RemainingWatchOnlyAccountCapacity returns the number of watch-only accounts that can be added.
func (api *API) RemainingWatchOnlyAccountCapacity(ctx context.Context) (int, error) {
	return (*api.messenger).RemainingWatchOnlyAccountCapacity()
}

// Creates all keystore files for a keypair and mark it in db as fully operable.
func (api *API) MakePrivateKeyKeypairFullyOperable(ctx context.Context, privateKey string, password string) error {
	return (*api.messenger).MakePrivateKeyKeypairFullyOperable(privateKey, password)
}

// CleanKeystoreFiles cleans the keystore files for all keypairs
// if the keypair is already migrated to keycard or removed, it cleans all accounts of the keypair, including the master account
// if the keypair is not migrated to keycard and not removed, it cleans the keystore files for removed accounts of the keypair,
// the master account is not cleaned if the keypair is not removed/migrated to keycard
func (api *API) CleanKeystoreFiles(ctx context.Context, password string) error {
	return api.manager.CleanKeystoreFiles(password)
}

func (api *API) MakePartiallyOperableAccoutsFullyOperable(ctx context.Context, password string) (addresses []types.Address, err error) {
	return api.manager.MakePartiallyOperableAccoutsFullyOperable(password)
}

// Creates all keystore files for a keypair and mark it in db as fully operable.
func (api *API) MakeSeedPhraseKeypairFullyOperable(ctx context.Context, mnemonic string, password string) error {
	mnemonicNoExtraSpaces := strings.Join(strings.Fields(mnemonic), " ")

	return (*api.messenger).MakeSeedPhraseKeypairFullyOperable(mnemonicNoExtraSpaces, password)
}

// Creates a random new mnemonic.
func (api *API) GetRandomMnemonic(ctx context.Context) (string, error) {
	return accscommon.CreateRandomMnemonicWithDefaultLength()
}

func (api *API) VerifyKeystoreFileForAccount(address types.Address, password string) bool {
	_, err := api.manager.LoadAccount(address, password)
	return err == nil
}

func (api *API) VerifyPassword(password string) bool {
	address, err := api.db.GetChatAddress()
	if err != nil {
		return false
	}
	return api.VerifyKeystoreFileForAccount(address, password)
}

func (api *API) MigrateNonProfileColdWalletKeypairToApp(ctx context.Context, mnemonic string, password string) error {
	mnemonicNoExtraSpaces := strings.Join(strings.Fields(mnemonic), " ")

	acc, err := generator.CreateAccountFromMnemonic(mnemonic, "")
	if err != nil {
		return err
	}

	keypair, err := api.db.GetKeypairByKeyUID(acc.KeyUID())
	if err != nil {
		return err
	}

	if keypair.Type == accsmanagementtypes.KeypairTypeProfile {
		return errors.New("cannot migrate profile keypair to app this way, use ConvertToRegularAccount instead")
	}

	return (*api.messenger).MigrateColdWalletKeypairToApp(ctx, mnemonicNoExtraSpaces, password)
}

func (api *API) MigrateNonProfileKeypairToColdWallet(ctx context.Context, keyUID string, password string, coldWallet accsmanagementtypes.ColdWalletType) error {
	keypair, err := api.db.GetKeypairByKeyUID(keyUID)
	if err != nil {
		return err
	}

	if keypair.Type == accsmanagementtypes.KeypairTypeProfile {
		return errors.New("cannot migrate profile keypair to cold wallet this way, use ConvertToKeycardAccount instead")
	}
	return (*api.messenger).MigrateKeypairToColdWallet(ctx, keyUID, password, coldWallet)
}

func (api *API) AddressWasShown(address types.Address) error {
	return api.db.AddressWasShown(address)
}

func (api *API) GetNumOfAddressesToGenerateForKeypair(keyUID string) (uint64, error) {
	return api.db.GetNumOfAddressesToGenerateForKeypair(keyUID)
}

func (api *API) ResolveSuggestedPathForKeypair(keyUID string) (string, error) {
	return api.db.ResolveSuggestedPathForKeypair(keyUID)
}

File: test/functional/clients/services/accounts.py
from clients.rpc import RpcClient
from clients.services.service import Service


class AccountService(Service):
    def __init__(self, client: RpcClient):
        super().__init__(client, "accounts")

    def get_accounts(self):
        response = self.rpc_request("getAccounts")
        return response

    def get_account_keypairs(self):
        response = self.rpc_request("getKeypairs")
        return response

    def add_account(self, password, account_data):
        params = [password, account_data]
        response = self.rpc_request("addAccount", params)
        return response

    def delete_account(self, account_address, password):
        params = [account_address, password]
        response = self.rpc_request("deleteAccount", params)
        return response

    def import_mnemonic(self, mnemonic, password):
        params = [mnemonic, password]
        response = self.rpc_request("importMnemonic", params)
        return response

    def add_keypair_via_seed_phrase(self, mnemonic, password, name, cold_wallet, wallet_account):
        params = [mnemonic, password, name, cold_wallet, wallet_account]
        response = self.rpc_request("addKeypairViaSeedPhrase", params)
        return response

    def add_keypair_via_private_key(self, private_key, password, name, wallet_account):
        params = [private_key, password, name, wallet_account]
        response = self.rpc_request("addKeypairViaPrivateKey", params)
        return response

    def verify_password(self, password):
        params = [password]
        response = self.rpc_request("verifyPassword", params)
        return response

    def resolve_suggested_path_for_keypair(self, key_uid):
        params = [key_uid]
        response = self.rpc_request("resolveSuggestedPathForKeypair", params)
        return response

    def has_paired_devices(self):
        response = self.rpc_request("hasPairedDevices", [])
        return response

    def update_keypair_name(self, key_uid, name):
        params = [key_uid, name]
        response = self.rpc_request("updateKeypairName", params)
        return response

    def move_wallet_account(self, from_position, to_position):
        params = [from_position, to_position]
        response = self.rpc_request("moveWalletAccount", params)
        return response

    def update_token_preferences(self, preferences):
        params = [preferences]
        response = self.rpc_request("updateTokenPreferences", params)
        return response

    def get_token_preferences(self):
        response = self.rpc_request("getTokenPreferences", [])
        return response

    def update_collectible_preferences(self, preferences):
        params = [preferences]
        response = self.rpc_request("updateCollectiblePreferences", params)
        return response

    def get_collectible_preferences(self):
        response = self.rpc_request("getCollectiblePreferences", [])
        return response

    def get_account_by_address(self, address):
        params = [address]
        response = self.rpc_request("getAccountByAddress", params)
        return response

    def get_keypair_by_key_uid(self, key_uid):
        params = [key_uid]
        response = self.rpc_request("getKeypairByKeyUID", params)
        return response

    def update_account(self, account):
        params = [account]
        response = self.rpc_request("updateAccount", params)
        return response

    def migrate_non_profile_keypair_to_cold_wallet(self, key_uid, password, cold_wallet):
        params = [key_uid, password, cold_wallet]
        response = self.rpc_request("migrateNonProfileKeypairToColdWallet", params)
        return response

    def migrate_non_profile_cold_wallet_keypair_to_app(self, mnemonic, password):
        params = [mnemonic, password]
        response = self.rpc_request("migrateNonProfileColdWalletKeypairToApp", params)
        return response

    def get_random_mnemonic(self):
        response = self.rpc_request("getRandomMnemonic", [])
        return response

    def add_keypair_stored_to_cold_wallet(self, key_uid, master_address, name, wallet_xpub, cold_wallet, wallet_accounts):
        params = [
            key_uid,
            master_address,
            name,
            wallet_xpub,
            cold_wallet,
            wallet_accounts,
        ]
        response = self.rpc_request("addKeypairStoredToColdWallet", params)
        return response

    def update_keypair(self, keypair):
        params = [keypair]
        response = self.rpc_request("updateKeypair", params)
        return response

    def get_watch_only_accounts(self):
        response = self.rpc_request("getWatchOnlyAccounts", [])
        return response

    def delete_keypair(self, key_uid, password):
        params = [key_uid, password]
        response = self.rpc_request("deleteKeypair", params)
        return response

    def remaining_account_capacity(self):
        response = self.rpc_request("remainingAccountCapacity", [])
        return response

    def remaining_keypair_capacity(self):
        response = self.rpc_request("remainingKeypairCapacity", [])
        return response

    def remaining_watch_only_account_capacity(self):
        response = self.rpc_request("remainingWatchOnlyAccountCapacity", [])
        return response

    def get_num_of_addresses_to_generate_for_keypair(self, key_uid):
        params = [key_uid]
        response = self.rpc_request("getNumOfAddressesToGenerateForKeypair", params)
        return response

    def verify_keystore_file_for_account(self, address, password):
        params = [address, password]
        response = self.rpc_request("verifyKeystoreFileForAccount", params)
        return response
